### PR TITLE
Add 'mean' and 'sum of squared deviations' stats to histogram

### PR DESCRIPTION
**Pull request by JayajP**
Add 'mean' and 'sum of squared deviations' stats to histogram. Need to record these stats since they are needed internally for EventMetrics (distributions). These statistics are updated using [Welford's method](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm) which is the standard online algorithm for updating mean/sum of squared deviations on the fly. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java [hunk 1]
```diff
@@ -184,6 +193,26 @@ public synchronized void record(double value) {
...
+  }
+
+  /**
+   * Update 'mean' and 'sum of squared deviations' statistics with the newly recorded value <a
+   * href="https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm">
+   * Welford Method</a>.
+   *
+   * @param value
+   */
+  private void updateStatistics(double value) {
```
Comment (m-trieu) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java:206]: nit: mark as synchronized to guard against a non-threadsafe/non synchronized caller

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java [hunk 2]
```diff
@@ -47,6 +47,9 @@ public class HistogramData implements Serializable {
   private long numTopRecords;
   private long numBottomRecords;
 
+  private double sum_of_squared_deviations;
```
Comment (m-trieu) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java:50]: mark as with `@GuardedBy("this")` unless there are non threadsafe callers

File: sdks/java/core/src/test/java/org/apache/beam/sdk/util/HistogramDataTest.java [hunk 3]
```diff
@@ -332,4 +332,19 @@ public void testExponentialBuckets_NumBuckets() {
     HistogramData negativeScaleBucket = HistogramData.exponential(-3, 500);
     assertThat(negativeScaleBucket.getBucketType().getNumBuckets(), equalTo(4));
   }
+
+  @Test
+  public void testStatistics() {
```
Comment (m-trieu) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/util/HistogramDataTest.java:337]: better to split this into 2 tests to test each part of the interface
testStatistics_mean
testStatistics_sumOfSquaredDeviations

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java [hunk 4]
```diff
@@ -58,6 +61,8 @@ public HistogramData(BucketType bucketType) {
     this.numBoundedBucketRecords = 0;
     this.numTopRecords = 0;
     this.numBottomRecords = 0;
+    this.mean = 0;
+    this.sum_of_squared_deviations = 0;
```
Comment (m-trieu) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java:65]: for java, camel case is preferred https://google.github.io/styleguide/javaguide.html#s5.3-camel-case

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java (hunk 2 above)
Comment (m-trieu) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java:50]: for java, camel case is preferred https://google.github.io/styleguide/javaguide.html#s5.3-camel-case

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java [hunk 5]
```diff
@@ -47,6 +47,9 @@ public class HistogramData implements Serializable {
   private long numTopRecords;
   private long numBottomRecords;
 
+  private double sum_of_squared_deviations;
+  private double mean;
```
Comment (m-trieu) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java:51]: mark as with @GuardedBy("this") unless there are non threadsafe callers

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java [hunk 6]
```diff
@@ -184,6 +193,26 @@ public synchronized void record(double value) {
...
+  private void updateStatistics(double value) {
+    long count = getTotalCount();
+    if (count == 1) {
+      mean = value;
+      return;
+    }
+
+    double old_mean = mean;
+    mean = old_mean + (value - old_mean) / count;
+    sum_of_squared_deviations += (value - mean) * (value - old_mean);
```
Comment (m-trieu) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java:215]: for java, camel case is preferred https://google.github.io/styleguide/javaguide.html#s5.3-camel-case

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java [hunk 7]
```diff
@@ -184,6 +193,26 @@ public synchronized void record(double value) {
...
+   * @param value
+   */
+  private void updateStatistics(double value) {
+    long count = getTotalCount();
+    if (count == 1) {
+      mean = value;
+      return;
+    }
+
+    double old_mean = mean;
```
Comment (m-trieu) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java:213]: for java, camel case is preferred https://google.github.io/styleguide/javaguide.html#s5.3-camel-case

File: sdks/java/core/src/test/java/org/apache/beam/sdk/util/HistogramDataTest.java [hunk 8]
```diff
@@ -332,4 +332,19 @@ public void testExponentialBuckets_NumBuckets() {
...
+    HistogramData histogram1 = HistogramData.linear(0, 10, 10);
+    for (int i = 0; i < 10; i++) {
+      histogram1.record(i * 10.0);
+    }
+
+    assertThat(histogram1.getMean(), equalTo(45.0));
+    double sum_of_squared_deviations = 0;
```
Comment (m-trieu) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/util/HistogramDataTest.java:344]: for java, camel case is preferred https://google.github.io/styleguide/javaguide.html#s5.3-camel-case

**Review by m-trieu — CHANGES_REQUESTED** [review]
mainly style and documentation

**Review by m-trieu — APPROVED** [review]
LGTM

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java (hunk 1 above)
Comment (JayajP) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java (hunk 2 above)
Comment (JayajP) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java]: done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/util/HistogramDataTest.java (hunk 3 above)
Comment (JayajP) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/util/HistogramDataTest.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java (hunk 2 above)
Comment (JayajP) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java (hunk 6 above)
Comment (JayajP) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java]: done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/util/HistogramDataTest.java (hunk 8 above)
Comment (JayajP) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/util/HistogramDataTest.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java (hunk 7 above)
Comment (JayajP) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java]: done

**Comment by JayajP** [comment]
One github actions test is failing for an unrelated reason. The corresponding jenkins test is passing.

**Review by johnjcasey — APPROVED** [review]

**Pull request merged**